extensions/eclipse-che-theia-plugin-ext/src/plugin/che-sidecar-file-system.ts

```diff
@@ -0,0 +1,223 @@
+
+import { RPCProtocol } from '@theia/plugin-ext/lib/common/rpc-protocol';
+import { URI } from 'vscode-uri';
+import { promisify } from 'util';
+
+export namespace SideCarFileSystemProvider {
+  export interface StatAndLink {
+    // The stats of the file. If the file is a symbolic
+    // link, the stats will be of that target file and
+    // not the link itself.
+    // If the file is a symbolic link pointing to a non
+    // existing file, the stat will be of the link and
+    // the `dangling` flag will indicate this.
+    stat: Stats;
+
+    // Will be provided if the resource is a symbolic link
+    // on disk. Use the `dangling` flag to find out if it
+    // points to a resource that does not exist on disk.
+    symbolicLink?: { dangling: boolean };
+  }
+}
+
+export enum FileSystemProviderErrorCode {
+  FileExists = 'EntryExists',
+  FileNotFound = 'EntryNotFound',
+  FileNotADirectory = 'EntryNotADirectory',
+  FileIsADirectory = 'EntryIsADirectory',
+  FileExceedsMemoryLimit = 'EntryExceedsMemoryLimit',
+  FileTooLarge = 'EntryTooLarge',
+  NoPermissions = 'NoPermissions',
+  Unavailable = 'Unavailable',
+  Unknown = 'Unknown',
+}
+
+export class FileSystemProviderError extends Error {
+  constructor(message: string, public readonly code: FileSystemProviderErrorCode) {
+    super(message);
+    Object.setPrototypeOf(this, FileSystemProviderError.prototype);
+  }
+}
+
+export function markAsFileSystemProviderError(error: Error, code: FileSystemProviderErrorCode): Error {
+  error.name = code ? `${code} (FileSystemError)` : 'FileSystemError';
+
+  return error;
+}
+
+export function createFileSystemProviderError(
+  error: Error | string,
+  code: FileSystemProviderErrorCode
+): FileSystemProviderError {
+  const providerError = new FileSystemProviderError(error.toString(), code);
+  markAsFileSystemProviderError(providerError, code);
+
+  return providerError;
+}
+
+export class CheSideCarFileSystemImpl implements CheSideCarFileSystem {
+  constructor(rpc: RPCProtocol) {
+    const delegate = rpc.getProxy(PLUGIN_RPC_CONTEXT.CHE_SIDECAR_FILE_SYSTEM_MAIN);
+    const machineName = process.env.CHE_MACHINE_NAME;
+    if (machineName) {
+      delegate.$registerFileSystemProvider(`file-sidecar-${machineName}`);
+    }
+  }
+
+  async $stat(resource: string): Promise<{ type: FileTypeMain; mtime: number; ctime: number; size: number }> {
+    try {
+      const statAndLink = await this.statLink(resource); // cannot use fs.stat() here to support links properly
+
+      return {
+        type: this.toType(statAndLink.stat, statAndLink.symbolicLink),
+        ctime: statAndLink.stat.birthtime.getTime(), // intentionally not using ctime here, we want the creation time
+        mtime: statAndLink.stat.mtime.getTime(),
+        size: statAndLink.stat.size,
+      };
+    } catch (error) {
+      throw this.toFileSystemProviderError(error);
+    }
+  }
+
+  $delete(resource: string, opts: { recursive: boolean; useTrash: boolean }): Promise<void> {
```

Comment:
If we don't implement an interface we define ourselves, let's remove the method from the interface.

Reply:
Agree. Methods that hasn't been implemented yet were removed.